keps/sig-node/3619-supplemental-groups-policy/README.md

```diff
@@ -0,0 +1,912 @@
+
+In Kubernetes API level, `PodSecurityContext.{RunAsUser, RunAsGroup, SupplementalGroups}` relates to this. This API was designed to override `config.User` configuration of OCI images. However, in the current implementation, as described in [kubernetes/kubernetes#112879](https://github.com/kubernetes/kubernetes/issues/112879), even when a manifest defines both `RunAsGroup`, __group memberships defined in the container image for the UID__ are attached to the container process (see the [the next section](#steps-to-reproduce) for details). This behavior clearly diverges from the specification of OCI image configuration, especially the next sentence of [`config.User` OCI image configuration](https://github.com/opencontainers/image-spec/blob/3a7f492d3f1bcada656a7d8c08f3f9bbd05e7406/config.md#:~:text=User%20string%2C%20OPTIONAL)):
+
+> If `group`/`gid` is not specified, the default group and supplementary groups of the given `user`/`uid` in `/etc/passwd` from the container are applied.
+
+As described in [kubernetes/kubernetes#112879](https://github.com/kubernetes/kubernetes/issues/112879), the behavior is not documented well and is not widely known by most Kubernetes administrators and users. Moreover, this behavior causes security considerations in some cases.
+
+### Steps to reproduce
+
+Assume you have an image and a Pod manifest:
+
+```Dockerfile
+# Dockerfile
+FROM ubuntu:22.04
+# This generates /etc/group entry --> "group-in-image:x:60000:alice"
+RUN groupadd -g 60000 group-in-image \
+    && useradd -m -u 1000 alice \
+    && gpasswd -a alice group-in-image
+USER alice
+```
+
+```yaml
+spec:
+  # This overrides 
+  # - USER directive in Dockerfile above by runAsUser and runAsGroup with "1000:1000", and 
+  # - setting supplementalGroups
+  # This spec expects NOT to attach gids defined in the image(/etc/group) to the container process
+  # because this specifies gid by runAsGroup explicitly.
+  securityContext: { runAsUser:1000, runAsGroup:1000, supplementalGroups:[60000]}
+  containers:
+    # Expected output: "uid=1000(alice) gid=1000(alice) groups=1000(alice),60000"
+    # NOTE: "group-in-image" is not included here 
+    #       because groups defined in /etc/group should not be attached
+    #       when gids is specified in runAsGroup
+  - image: the-image-above
+    sh: ["id"] 
+```
+
+However, the current combination with Kubernetes and major containter runtimes(at least containerd and cri-o) outputs(See [here](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime/tree/reproduce-bypass-supplementalgroups) for more detailed reproduction code) includes "group-in-image" group of the first container process.
+
+```console
+uid=1000(alice) gid=1000(alice) groups=1000(alice),50000(group-in-image),60000
```

Comment:
What is 50000?

Reply:
just an example.

`/etc/group` in the image contains the line:

```
group-in-image:x:60000:alice
```